fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+            WireId wire = ctx->getPipDstWire(pip);
+
+            if (wire == WireId())
+                continue;
+
+            ChainWireNode next_node;
+            next_node.wire = wire;
+            next_node.depth = node_to_expand.depth += 1;
+
+            if (next_node.depth >= 5) {
+                break;
+            }
+
+            auto const &wire_data = ctx->wire_info(wire);
+
+            bool expand_node = true;
+            if (ctx->is_site_port(pip)) {
+                switch (node_to_expand.state) {
+                case IN_SOURCE_SITE:
+                    NPNR_ASSERT(wire_data.site == -1);
+                    next_node.state = IN_ROUTING;
+                    break;
+                case IN_ROUTING:
+                    NPNR_ASSERT(wire_data.site != -1);
+                    next_node.state = IN_SINK_SITE;
+                    break;
+                case IN_SINK_SITE:
+                    expand_node = false;
+                    break;
+                default:
+                    // Unreachable!!!
+                    NPNR_ASSERT(false);
+                }
+            } else {
+                next_node.state = node_to_expand.state;
+            }
+
+            if (expand_node) {
+                nodes_to_expand.push_back(next_node);
+            } else {
+                continue;
+            }
+
+            if (next_node.state == IN_SINK_SITE) {
+                for (BelPin bel_pin : ctx->getWireBelPins(wire)) {
+                    BelId sink_bel = bel_pin.bel;
+                    auto const &bel_data = bel_info(ctx->chip_info, sink_bel);
+
+                    if (bel_data.category != BEL_CATEGORY_LOGIC)
+                        continue;
+
+                    if (bel_data.synthetic)
+                        continue;
+
+                    if (ctx->getBelPinType(bel_pin.bel, bel_pin.pin) != PORT_IN)
+                        continue;
+
+                    if (sink_bel.index == bel.index)
+                        return sink_bel;
+                }
+            }
+        }
+    }
+
+    BelId tmp_bel;
+    return tmp_bel;
+}
+
+CellInfo* Arch::getClusterRootCell(ClusterId cluster) const
+{
+    NPNR_ASSERT(cluster != ClusterId());
+    CellInfo *ci = nullptr;
+    auto cell = getCtx()->cells.find(cluster);
+    if (cell != getCtx()->cells.end()) {
+        ci = cell->second.get();
+    }
+    return ci;
+}
+
+bool Arch::getClusterPlacement(ClusterId cluster, BelId root_bel,
+                         std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    const Context *ctx = getCtx();
+
+    // Place root
+    CellInfo *root_cell = getClusterRootCell(cluster);
+
+    if (!ctx->isValidBelForCellType(root_cell->type, root_bel))
+        return false;
+
+    // Place cluster
+    // Prepare first loop iteration
+    CellInfo *next_cell = root_cell;
+    NetInfo *next_net = nullptr;
+    BelId next_bel = root_bel;
+    Loc next_loc = getBelLocation(next_bel);
+
+    std::vector<IdString> placed_bels;
```

Comment:
I think using `pool` (i.e.  the new `unordered_set` replacement) would make more sense than `std::vector`, as far as I can tell the only thing you do with this structure is check membership (which can be done significantly faster with a hash set than `O(n)` in a vector) with nothing that cares about ordering